### PR TITLE
Add support for matching <firmware> requirements on device parents

### PR DESCRIPTION
**Pull request by hughsie**
For composite devices you might want to restrict the child device in respect
to the parent, for instance requiring the parent to have greater than a specific
version number.

The other useful thing to use this for is checking if the parent has a specific
GUID (of any version) which allows us to match against the common VID&PID
instance IDs. This would allow us to restrict a generic child device update to
a specific OEM vendor parent.

This is specified as <firmware depth="1" ...> to match the parent device and
<firmware depth="2" ...> to match the grandparent device.

Type of pull request:
- [ ] New plugin (Please include [new plugin checklist](https://github.com/hughsie/fwupd/wiki/New-plugin-checklist))
- [ ] Code fix
- [ ] Feature
- [ ] Documentation

### CODE REVIEW

File: src/fu-engine.c [hunk 1]
```diff
@@ -1079,11 +1079,29 @@ static gboolean
...
+	/* look at the parent device */
+	depth = xb_node_get_attr_as_uint (req, "depth");
+	if (depth != G_MAXUINT64) {
+		for (guint i = 0; i < depth; i++) {
+			device_actual = fu_device_get_parent (device_actual);
+			if (device_actual == NULL) {
+				g_set_error (error,
+					     FWUPD_ERROR,
+					     FWUPD_ERROR_NOT_SUPPORTED,
+					     "No parent device for %s (%u/%" G_GUINT64_FORMAT ")",
```
Comment (superm1) [inline src/fu-engine.c]: I personally think that name is more useful than device id in an error message.

File: src/fu-engine.c [hunk 2]
```diff
@@ -1079,11 +1079,29 @@ static gboolean
 fu_engine_check_requirement_firmware (FuEngine *self, XbNode *req,
 				      FuDevice *device, GError **error)
 {
+	FuDevice *device_actual = device;
+	guint64 depth;
 	g_autoptr(GError) error_local = NULL;
 
+	/* look at the parent device */
+	depth = xb_node_get_attr_as_uint (req, "depth");
+	if (depth != G_MAXUINT64) {
```
Comment (superm1) [inline src/fu-engine.c]: Is this the default return if unset or the extreme? If it's the extreme I think this should be an error condition that it's this value.

File: src/fu-engine.c [hunk 3]
```diff
@@ -1079,11 +1079,29 @@ static gboolean
...
+		for (guint i = 0; i < depth; i++) {
```
Comment (superm1) [inline src/fu-engine.c]: Doesn't this need to be guint64 to make sure you compare same types?

File: src/fu-engine.c (hunk 2 above)
Comment (hughsie) [reply src/fu-engine.c]: It's the value for unset.

File: src/fu-engine.c (hunk 3 above)
Comment (hughsie) [reply src/fu-engine.c]: Good catch, although I think it's unlikely to have more than 4294967295 nested children :)

File: src/fu-engine.c (hunk 1 above)
Comment (hughsie) [reply src/fu-engine.c]: Sure, that works too. Fixed.

File: src/fu-engine.c (hunk 3 above)
Comment (superm1) [reply src/fu-engine.c]: I sure hope not 😆

**Comment by hughsie** [comment]
Actually, this needs more work to do the GUID matching.

**Comment by hughsie** [comment]
Okay, this one should work. Self tests FTW.

File: src/fu-engine.c [hunk 4]
```diff
@@ -1149,37 +1169,50 @@ fu_engine_check_requirement_firmware (FuEngine *self, XbNode *req,
...
 		const gchar *guid = xb_node_get_text (req);
 		const gchar *version;
-		g_autoptr(FuDevice) device2 = NULL;
 
 		/* find if the other device exists */
-		device2 = fu_device_list_get_by_guid (self->device_list, guid, error);
-		if (device2 == NULL)
-			return FALSE;
+		if (depth == G_MAXUINT64) {
+			device_actual = fu_device_list_get_by_guid (self->device_list, guid, error);
```
Comment (superm1) [inline src/fu-engine.c]: doesn't this need to be a g_set_object call to since you took a reference above?

Comment (hughsie) [reply src/fu-engine.c]: Good catch! Fixed.

File: src/fu-engine.c [hunk 5]
```diff
@@ -1079,11 +1079,31 @@ static gboolean
...
 {
+	guint64 depth;
+	g_autoptr(FuDevice) device_actual = g_object_ref (device);
 	g_autoptr(GError) error_local = NULL;
 
+	/* look at the parent device */
+	depth = xb_node_get_attr_as_uint (req, "depth");
+	if (depth != G_MAXUINT64) {
+		for (guint64 i = 0; i < depth; i++) {
+			FuDevice *device_tmp = fu_device_get_parent (device_actual);
```
Comment (superm1) [inline src/fu-engine.c]: I think this might need to be `g_autoptr` and take a reference too since you're assigning it to another `g_autoptr`

Comment (hughsie) [reply src/fu-engine.c]: Nah, `fu_device_get_parent()` is `transfer-none`

Comment (superm1) [reply src/fu-engine.c]: But then you do a `g_set_object` later, it's still OK?  I'm worried on cleanup you destroy the parent.

Comment (hughsie) [reply src/fu-engine.c]: `g_set_object()` clears the first parameter if non-NULL, then refs the second parameter, saving it into the first. It's like doing:

    if (device_actual != NULL)
        g_object_unref (device_actual);
    device_actual = g_object_ref (device_tmp)

...but as one atomic thing.

Comment (superm1) [reply src/fu-engine.c]: Ah yeah - thanks for clarifying.

**Pull request merged**